testing/src/scenario/state.py

```diff
@@ -2055,7 +2055,11 @@ def _is_builtin_event(self, charm_spec: _CharmSpec[CharmType]) -> bool:
         return self._path.type is not _EventType.CUSTOM
 
     def deferred(self, handler: Callable[..., Any], event_id: int = 1) -> DeferredEvent:
-        """Construct a DeferredEvent from this Event."""
+        """Construct a DeferredEvent from this Event.
+
+        The charm class passed to the :class:`Context` must be observing the
+        event this DeferredEvent is based on.
```

Comment:
I believe this is a new requirement (a test that did not do this started failing after the changes). However, I feel that it should always have been the case, and if charms are dynamically observing events, then they need to have static versions of those for the unit tests.

Without this restriction, the type doesn't get registered.

Reply:
Do we wipe deferred events on charm upgrade?
I imagine that the unit state db is kept during the upgrade of vm charms.
In which case there's no hard guarantee.

Reply:
> Do we wipe deferred events on charm upgrade?

It depends - on K8s the unit state DB is lost because you get a new container, but on machines it's in-place so they stay.

> In which case there's no hard guarantee.

I still feel that the charm code should handle this situation. But in any case, as per the comment in David's PR that took this fix, it should be handled in a separate issue if anyone does need this functionality. It's not really related to this PR or to that one, just triggered by it, and it's not what the test is intended to check.